### ISSUE TITLE
lirc: not works

### ISSUE THREAD

**Issue by nE0sIghT** [nE0sIghT/android_kernel_xiaomi_msm8953#20]
4.9 latest commit.

dmesg:
```
[    2.562346] lirc_dev: IR Remote Control driver registered, major 229
[    2.562371] IR LIRC bridge handler initialized
[    2.562377] peelir_init:npages = 150
[    4.614650] pwm-ir soc:pwm_ir: reg-id = vdd, low-active = 0, use-timer = 0
[    4.614769] Registered IR keymap rc-lirc
[    4.614868] input: pwm-ir as /devices/platform/soc/soc:pwm_ir/rc/rc0/input5
[    4.614907] rc rc0: pwm-ir as /devices/platform/soc/soc:pwm_ir/rc/rc0
[    4.615183] rc rc0: lirc_dev: driver ir-lirc-codec (pwm-ir) registered at minor = 0
```

logcat:
```
04-30 09:17:04.740   427   427 I ServiceManager: service 'consumer_ir' died
04-30 09:17:11.134  2385  2607 D EventHub: No input device configuration file found for device 'pwm-ir'.
```

**Comment by nE0sIghT** [comment]
3.18 dmesg looks same:

```
[    1.812150] lirc_dev: IR Remote Control driver registered, major 233 
[    5.366175] pwm-ir soc:pwm_ir: reg-id = vdd, low-active = 0, use-timer = 0
[    5.366456] Registered IR keymap rc-lirc
[    5.366614] input: pwm-ir as /devices/soc/soc:pwm_ir/rc/rc0/input5
[    5.366672] rc0: pwm-ir as /devices/soc/soc:pwm_ir/rc/rc0
[    5.366814] rc rc0: lirc_dev: driver ir-lirc-codec (pwm-ir) registered at minor = 0
```

logcat also contains to lines:
```
04-28 16:51:10.284   445   445 I ServiceManager: service 'consumer_ir' died
04-28 20:51:23.328  2435  2668 D EventHub: No input device configuration file found for device 'pwm-ir'.
```

Anyone else can test lirc?

**Comment by nE0sIghT** [comment]
Debug enabled, dmesg while sending IR signals:

```
[  128.485274] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  129.288974] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  130.093897] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  130.900916] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  131.706528] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  132.513076] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  133.326324] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  134.126616] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  134.932838] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  135.738466] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  136.548182] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  137.355972] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  138.164614] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  138.971863] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  139.780850] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  140.588371] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called

```

**Comment by nE0sIghT** [comment]
Debug branch: https://github.com/nE0sIghT/android_kernel_xiaomi_msm8953/tree/lineage-16.0-mido-4.9-lirc

dmesg log while sending IR signal:
```
[  132.876571] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): open called
[  132.876591] rc rc0: ir_lirc_open
[  132.876597] rc rc0: ir_lirc_open ret: 0
[  132.876605] rc rc0: lirc_dev (ir-lirc-codec (pwm-ir)[0]): error: 0
[  132.876755] ir_lirc_transmit_ir
[  132.876761] ir_lirc_transmit_ir: lirc_get_pdata
[  132.876764] ir_lirc_transmit_ir: sizeof
[  132.876772] rc rc0: ir_lirc_transmit_ir: lirc->dev
[  132.876777] rc rc0: ir_lirc_transmit_ir: dev->tx_ir
[  132.876783] rc rc0: pwm_ir_tx_transmit
[  132.876793] rc rc0: regulator_enable
[  132.876797] pwm_ir_tx_transmit_with_delay enter
[  132.876992] pwm_ir_tx_work enter
[  133.130265] pwm_ir_tx_work out
[  133.130320] pwm_ir_tx_transmit_with_delay out
[  133.130341] rc rc0: regulator_disable
[  133.130347] rc rc0: ir_lirc_transmit_ir: dev->tx_ir
[  133.130351] rc rc0: ir_lirc_transmit_ir: out
[  133.131547] rc rc0: ir_lirc_close
```

It should work...

**Comment by nE0sIghT** [comment]
Mainline kernel got own implementation of pwm ir: https://github.com/torvalds/linux/blob/master/drivers/media/rc/pwm-ir-tx.c

**Comment by nE0sIghT** [comment]
Backported mainline driver don't works as well.  
Looks like some dts problem - maybe something prevents powering ir?

**Comment by Flex1911** [comment]
It's very likely that you need to change the sysfs path and/or permissions in userspace, since 4.9 paths have changed a bit.

**Comment by nE0sIghT** [comment]
IR device is accessed by ```/dev/lirc0```.  
The whole process is:
/dev/lirc0 → media/rc/lirc_dev → media/rc/pwm-ir → pwm/pwm-qpnp

I traced it to pwm-qpnp - all registers are writen well.

**Comment by nE0sIghT** [comment]
dmesg while sending IR signal: https://gist.github.com/nE0sIghT/04ff5040a95faf5d57a7f5d59bfbdba8

Perfectly ok, but not works.

**Comment by nE0sIghT** [comment]
I done simple IR camera test.
With 3.18 diode lights up: https://www.youtube.com/watch?v=VHa7vrf5ImE
With 4.9 it not lights up despite successfull logs. So something prevents diode powering.

**Comment by nE0sIghT** [comment]
```
[    4.617150] pwm-ir soc:pwm_ir: reg-id = vdd, low-active = 0, use-timer = 0
[    4.617224] vdd requested: /soc/pwm_ir, exclusive: no
[    4.617228] vdd regulator: soc:pwm_ir-vdd, min_uV: 0, max_uV: 0
```

**Comment by nE0sIghT** [comment]
4.9 while sending IR signal:
```
mido:/sys/kernel/debug/regulator/soc:qcom,rpm-smd:rpm-regulator-ldoa8:regulator-l8-pm8953_l8 # cat consumers                                                                             
Device-Supply                    EN    Min_uV   Max_uV  load_uA
soc:pwm_ir-vdd                   Y    2900000  2900000        0
7824900.sdhci-vdd                Y    2900000  2900000   570000
pm8953_l8                        N          0        0        0
```

**Comment by xCuri0** [comment]
> 4.9 while sending IR signal:
> 
> ```
> mido:/sys/kernel/debug/regulator/soc:qcom,rpm-smd:rpm-regulator-ldoa8:regulator-l8-pm8953_l8 # cat consumers                                                                             
> Device-Supply                    EN    Min_uV   Max_uV  load_uA
> soc:pwm_ir-vdd                   Y    2900000  2900000        0
> 7824900.sdhci-vdd                Y    2900000  2900000   570000
> pm8953_l8                        N          0        0        0
> ```

I see 0uV on the regulator while it shows enabled and I'm trying IR. Could the problem be a change in kernel 4.9 which makes other consumers unable to draw power from regulator sdhci is on ?

**Comment by nE0sIghT** [comment]
> I see 0uV on the regulator

You can strip regulator stuff from ir dts - it doesn't needed after all.

**Comment by xCuri0** [comment]
@nE0sIghT I did some minor changes to pwm-ir to get it sending the exact same stuff through SPMI
```
diff --git a/drivers/media/rc/pwm-ir.c b/drivers/media/rc/pwm-ir.c
index 0b006a35cbd4..380458891bfe 100644
--- a/drivers/media/rc/pwm-ir.c
+++ b/drivers/media/rc/pwm-ir.c
@@ -47,6 +47,30 @@ struct pwm_ir_packet {
 #define __devinit
 #define __devexit_p
 
+int mi_pwm_config(struct pwm_device *pwm, int duty_ns, int period_ns)
+{
+       int rc;
+       struct pwm_state state;
+
+       if (!pwm)
+               return -EINVAL;
+
+       if (duty_ns < 0 || period_ns < 0)
+               return -EINVAL;
+
+       pwm_get_state(pwm, &state);
+
+       state.duty_cycle = duty_ns;
+       state.period = period_ns;
+       rc = pwm->chip->ops->config(pwm->chip, pwm,
+                                       state.duty_cycle,
+                                       state.period);
+       pwm->state.duty_cycle = state.duty_cycle;
+       pwm->state.period = state.period;
+
+       return rc;
+}
+
 /* code for ir transmit */
 static int pwm_ir_tx_config(struct pwm_ir_dev *dev, u32 carrier, u32 duty_cycle)
 {
@@ -55,7 +79,7 @@ static int pwm_ir_tx_config(struct pwm_ir_dev *dev, u32 carrier, u32 duty_cycle)
        period_ns = NSEC_PER_SEC / carrier;
        duty_ns = period_ns * duty_cycle / 100;
 
-       rc = pwm_config(dev->pwm, duty_ns, period_ns);
+       rc = mi_pwm_config(dev->pwm, duty_ns, period_ns);
        if (rc == 0) {
                dev->carrier = carrier;
                dev->duty_cycle = duty_cycle;
```
Still doesn't work though. If regulators aren't the problem I guess something has changed between the old 3.18 msm_spmi driver and the new linux spmi driver

**Comment by xCuri0** [comment]
Confirmed it is the SPMI driver. I tried to fix it but couldn't

wrbase on 3.18 is 14680064
wrbase on 4.9 is 209715200

maybe some other part requires it be mapped at the 3.18 address for it to work

[pmicwrite3.18.log](https://github.com/nE0sIghT/android_kernel_xiaomi_msm8953/files/5067838/pmicwrite3.18.log)
[pmicwrite4.9.log](https://github.com/nE0sIghT/android_kernel_xiaomi_msm8953/files/5067839/pmicwrite4.9.log)

**Comment by nE0sIghT** [comment]
Fixed in 447867a8
